rts/idris_rts.c

```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
+    // this is an overapproximation of what's needed to hold an int as a string
+    // Each byte will take log_10(256) ~= 2.4 digits, so 4 digits per byte is
+    // clearly enough. Then add 1 for the terminator.
+    char temp[sizeof(int) * 4 + 1] = {0};
     int x = (int) GETINT(i);
-    Closure* cl = allocate(sizeof(Closure) + sizeof(char)*16, 0);
-    SETTY(cl, STRING);
-    cl -> info.str = (char*)cl + sizeof(Closure);
-    sprintf(cl -> info.str, "%d", x);
-    return cl;
+    sprintf(temp, "%d", x);
+
+    // avoid invalid UTF-8: all ASCII digits are acceptable as UTF-8 characters.
+    unsigned int j;
+    for (j = 0; j < sizeof(int) * 4 + 1; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j])) && temp[j] != '-') {
+            fprintf(stderr, "Fatal error: invalid output from sprintf on an int!\n");
+            fprintf(stderr, "It was [%c] at index [%u] in [%s].\n", temp[j], j, temp);
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    return MKSTR(vm, (const utf8_byte*)temp);
 }
 
 VAL idris_castBitsStr(VM* vm, VAL i) {
-    Closure* cl;
     ClosureType ty = i->ty;
+    // max length 64 bit unsigned int str 20 chars (18,446,744,073,709,551,615)
+    char temp[21] = { 0 };
 
     switch (ty) {
     case BITS8:
-        // max length 8 bit unsigned int str 3 chars (256)
-        cl = allocate(sizeof(Closure) + sizeof(char)*4, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu8, (uint8_t)i->info.bits8);
+        sprintf(temp, "%" PRIu8, (uint8_t)i->info.bits8);
         break;
     case BITS16:
-        // max length 16 bit unsigned int str 5 chars (65,535)
-        cl = allocate(sizeof(Closure) + sizeof(char)*6, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu16, (uint16_t)i->info.bits16);
+        sprintf(temp, "%" PRIu16, (uint16_t)i->info.bits16);
         break;
     case BITS32:
-        // max length 32 bit unsigned int str 10 chars (4,294,967,295)
-        cl = allocate(sizeof(Closure) + sizeof(char)*11, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu32, (uint32_t)i->info.bits32);
+        sprintf(temp, "%" PRIu32, (uint32_t)i->info.bits32);
         break;
     case BITS64:
-        // max length 64 bit unsigned int str 20 chars (18,446,744,073,709,551,615)
-        cl = allocate(sizeof(Closure) + sizeof(char)*21, 0);
-        cl->info.str = (char*)cl + sizeof(Closure);
-        sprintf(cl->info.str, "%" PRIu64, (uint64_t)i->info.bits64);
+        sprintf(temp, "%" PRIu64, (uint64_t)i->info.bits64);
         break;
     default:
         fprintf(stderr, "Fatal Error: ClosureType %d, not an integer type", ty);
         exit(EXIT_FAILURE);
     }
 
-    SETTY(cl, STRING);
-    return cl;
+    // avoid invalid UTF-8: all ASCII digits are acceptable as UTF-8 characters.
+    int j;
+    for (j = 0; j < 21; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j]))) {
```

Comment:
I came to the same conclusion as @melted -- we do similar validation in about 4 places IIRC, so it is definitely worth factoring out. It is ok if the set of characters accepted is an overapproximation as long as it is a subset of US-ASCII.